### ISSUE TITLE
Unable to install plugins on Jenkins master

### ISSUE THREAD

**Issue by sernapallyanurag** [chef-cookbooks/jenkins#469]
### Cookbook version

Jenkins-2.4.1
### Chef-client version

12.9.38
### Platform Details

Ubuntu 14.04-LTS
### Scenario:

I am trying to install plugins on Jenkins using Jenkins resource "jenkins_plugin"
### Steps to Reproduce:

I am using .hpi method to install plugins.
### Expected Result:

Installation of plugins ( I have mentioned plugin dependency url before installation of target plugin)
### Actual Result:

``

Recipe: jenkins::nodejs_plugin  \* jenkins_plugin[nodejs] action install
  Recipe: <Dynamically Defined Resource>
    \* remote_file[/var/chef/cache/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin] action create[2016-04-21T14:35:24+00:00] ERROR: Timeout connecting to https://updates.jenkins-ci.org/download/plugins/nodejs/0.2.1/nodejs.hpi, retry 1/5
[2016-04-21T14:37:36+00:00] ERROR: Error connecting to https://updates.jenkins-ci.org/download/plugins/nodejs/0.2.1/nodejs.hpi, retry 2/5
[2016-04-21T14:39:49+00:00] ERROR: Error connecting to https://updates.jenkins-ci.org/download/plugins/nodejs/0.2.1/nodejs.hpi, retry 3/5
 (up to date)
    \* remote_file[/var/chef/cache/jenkins-cli.jar] action create (up to date)
    \* remote_file[/var/chef/cache/update-center.json] action create (up to date)
    \* file[/var/chef/cache/extracted-update-center.json] action create (up to date)

```
================================================================================
Error executing action `install` on resource 'jenkins_plugin[nodejs]'
================================================================================

Mixlib::ShellOut::ShellCommandFailed
------------------------------------
Expected process to exit with [0], but received '255'
---- Begin output of "java" -jar "/var/chef/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin/var/chef/cache/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin -name nodejs  ----
STDOUT:
STDERR: java.io.IOException: No X-Jenkins-CLI2-Port among [null, X-Required-Permission, X-Jenkins, X-You-Are-In-Group, X-Hudson, Content-Length, Expires, X-You-Are-Authenticated-As, X-Permission-Implied-By, Set-Cookie, Server, X-Content-Type-Options, Date, X-Jenkins-Session, Content-Type]
    at hudson.cli.CLI.getCliTcpPort(CLI.java:290)
    at hudson.cli.CLI.<init>(CLI.java:128)
    at hudson.cli.CLIConnectionFactory.connect(CLIConnectionFactory.java:72)
    at hudson.cli.CLI._main(CLI.java:479)
    at hudson.cli.CLI.main(CLI.java:390)
    Suppressed: java.io.IOException: Server returned HTTP response code: 403 for URL: http://localhost:8080/cli
            at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1628)
            at hudson.cli.FullDuplexHttpStream.<init>(FullDuplexHttpStream.java:78)
            at hudson.cli.CLI.connectViaHttp(CLI.java:158)
            at hudson.cli.CLI.<init>(CLI.java:132)
            ... 3 more
---- End output of "java" -jar "/var/chef/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin/var/chef/cache/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin -name nodejs  ----
Ran "java" -jar "/var/chef/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /var/chef/cache/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin -name nodejs  returned 255

Cookbook Trace:
---------------
/var/chef/cache/cookbooks/jenkins/libraries/_executor.rb:84:in `execute!'
/var/chef/cache/cookbooks/jenkins/libraries/plugin.rb:316:in `install_plugin_from_url'
/var/chef/cache/cookbooks/jenkins/libraries/plugin.rb:116:in `block (2 levels) in <class:JenkinsPlugin>'
/var/chef/cache/cookbooks/jenkins/libraries/plugin.rb:153:in `block in <class:JenkinsPlugin>'

Resource Declaration:
---------------------
# In /var/chef/cache/cookbooks/jenkins/recipes/nodejs_plugin.rb

  3: jenkins_plugin 'nodejs' do
  4:   source 'https://updates.jenkins-ci.org/latest/nodejs.hpi'
  5:   action :install
  6:   notifies :restart, 'service[jenkins]', :immediately
  7: end

Compiled Resource:
------------------
# Declared in /var/chef/cache/cookbooks/jenkins/recipes/nodejs_plugin.rb:3:in `from_file'

jenkins_plugin("nodejs") do
  action [:install]
  retries 0
  retry_delay 2
  default_guard_interpreter :default
  declared_type :jenkins_plugin
  cookbook_name "jenkins"
  recipe_name "nodejs_plugin"
  source "https://updates.jenkins-ci.org/latest/nodejs.hpi"
end

Platform:
---------
x86_64-linux
```

Running handlers:
[2016-04-21T14:40:10+00:00] ERROR: Running exception handlers
Running handlers complete
[2016-04-21T14:40:10+00:00] ERROR: Exception handlers complete
Chef Client failed. 4 resources updated in 11 minutes 19 seconds
[2016-04-21T14:40:11+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2016-04-21T14:40:11+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2016-04-21T14:40:11+00:00] ERROR: jenkins_plugin[nodejs](jenkins::nodejs_plugin line 3) had an error: Mixlib::ShellOut
::ShellCommandFailed: Expected process to exit with [0], but received '255'
---- Begin output of "java" -jar "/var/chef/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /var/chef/cac
he/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin -name nodejs  ----
STDOUT:
STDERR: java.io.IOException: No X-Jenkins-CLI2-Port among [null, X-Required-Permission, X-Jenkins, X-You-Are-In-Group, X
-Hudson, Content-Length, Expires, X-You-Are-Authenticated-As, X-Permission-Implied-By, Set-Cookie, Server, X-Content-Type-Options, Date, X-Jenkins-Session, Content-Type]
        at hudson.cli.CLI.getCliTcpPort(CLI.java:290)
        at hudson.cli.CLI.<init>(CLI.java:128)
        at hudson.cli.CLIConnectionFactory.connect(CLIConnectionFactory.java:72)
        at hudson.cli.CLI._main(CLI.java:479)
        at hudson.cli.CLI.main(CLI.java:390)
        Suppressed: java.io.IOException: Server returned HTTP response code: 403 for URL: http://localhost:8080/cli
                at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1628)
                at hudson.cli.FullDuplexHttpStream.<init>(FullDuplexHttpStream.java:78)
                at hudson.cli.CLI.connectViaHttp(CLI.java:158)
                at hudson.cli.CLI.<init>(CLI.java:132)
                ... 3 more
---- End output of "java" -jar "/var/chef/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /var/chef/cache
/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin -name nodejs  ----
Ran "java" -jar "/var/chef/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /var/chef/cache/nodejs-9db830c2e86512008cb5adfff203ec3c.plugin -name nodejs  returned 255
[2016-04-21T14:40:12+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
``

**Comment by pmoosh** [comment]
Any change you just installed jenkins 2.0?

**Comment by IevgenKabanets** [comment]
Jenkins 2.0 is being installed by default. And it has auth by default.
Problem is - following attribute is not working, while installing from package (not present in apt repo).
`node['jenkins']['master']['version'] = "***"`

**Comment by sernapallyanurag** [comment]
Yes its installing version 2.0 by default. How can we revert it back to previous version to make Jenkins resources work.

**Comment by daften** [comment]
put the actual version in the attribute.
Also, duplicates #466

**Comment by IevgenKabanets** [comment]
@daften what do U mean "put"?
Obviously I just put there *** for example, as neither one of versions U put there - will work. 
Apt repo has no candidates for any 1.*** version.

**Comment by jsyrjala** [comment]
I fixed this by doing this in a recipe (Ubuntu Linux).

```
node.set[:jenkins][:master][:repository] = 'http://pkg.jenkins-ci.org/debian-stable'
node.set[:jenkins][:master][:repository_key] = 'http://pkg.jenkins-ci.org/debian-stable/jenkins-ci.org.key'
```

This will fetch jenkins package from debian-stable repository that contains Jenkins LTS releases. LTS releases are in 1.6.x branch at least for some time.

**Comment by sernapallyanurag** [comment]
I can't even install jenkins now.
``
      ================================================================================
      Error executing action `create` on resource 'remote_file[/var/chef/cache/jenkins-ci.org.key]'
      ================================================================================

      Errno::ECONNREFUSED
      -------------------
      Connection refused - Connection refused connecting to http://pkg.jenkins-ci.org/debian/jenkins-ci.org.key, giving
up

      Resource Declaration:
      ---------------------
      # In /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-2.9.38/lib/chef/provider/apt_repository.rb

      170:         declare_resource(type, cached_keyfile) do
      171:           source new_resource.key
      172:           mode "0644"
      173:           sensitive new_resource.sensitive
      174:           action :create
      175:         end
      176:

      Compiled Resource:
      ------------------
      # Declared in /opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38/lib/chef/provider/apt_repository.rb:170:in`install_key_from_uri'

      remote_file("/var/chef/cache/jenkins-ci.org.key") do
        provider Chef::Provider::RemoteFile
        action [:create]
        retries 0
        retry_delay 2
        default_guard_interpreter :default
        source ["http://pkg.jenkins-ci.org/debian/jenkins-ci.org.key"]
        use_etag true
        use_last_modified true
        declared_type :remote_file
        cookbook_name "jenkins"
        mode "0644"
        atomic_update true
        path "/var/chef/cache/jenkins-ci.org.key"
      end

      Platform:
      ---------
      x86_64-linux


    ================================================================================
    Error executing action `add` on resource 'apt_repository[jenkins]'
    ================================================================================

    Errno::ECONNREFUSED
    -------------------
    remote_file[/var/chef/cache/jenkins-ci.org.key] (/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.9.38/lib/chef/p
rovider/apt_repository.rb line 170) had an error: Errno::ECONNREFUSED: Connection refused - Connection refused connecting to http://pkg.jenkins-ci.org/debian/jenkins-ci.org.key, giving up

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/jenkins/recipes/_master_package.rb
``

**Comment by daften** [comment]
@IevgenKabanets  I mean, put this in the attributes of the cookbook; role; environment; ... (whatever suits your way or working)
```
override['jenkins']['master']['version'] = '1.658-1.1'
```
The solution from @jsyrjala should indeed work for a limited time, see also https://github.com/chef-cookbooks/jenkins/blob/master/attributes/master.rb#L195

I've opened a PR that should enable jenkins2 support and would like feedback there. There are still some manual steps needed, and the checks are too limited, but I got it too work now at least.

**Comment by IevgenKabanets** [comment]
@daften yes, I know how to set attribs for chef, but obviously it didn't work, because of the apt repo.
Will check the PR, thnx.

**Comment by daften** [comment]
@IevgenKabanets  Just use my fix, without changing the apt-repo, and that should work. Or check if the version number is correct, the version number I gave is for centOS ;)

**Comment by IevgenKabanets** [comment]
@daften 
It won't work, I'm afraid. Here's why:

```
# apt-cache policy jenkins
jenkins:
  Installed: 1.651.1
  Candidate: 2.1
  Version table:
     2.1 0
        500 http://pkg.jenkins-ci.org/debian/ binary/ Packages
 *** 1.651.1 0
        100 /var/lib/dpkg/status
```
Even with `override['jenkins']['master']['version'] = '1.658'`. If apt-repo offers 2.1 (while 1.658 is [there ](http://pkg.jenkins-ci.org/debian/binary/jenkins_1.658_all.deb)too). 

Using fix from @jsyrjala - will install 1.651.1 (which is not latest of 1.*** versions).

**Comment by daften** [comment]
That is because you're using the stable repo i'm guessing. That doesn't have 1.658. Don't just blindly try to combine solutions ;)

**Comment by IevgenKabanets** [comment]
@daften 
I'm using 'latest' repo. This one - http://pkg.jenkins-ci.org/debian/
With override attribute (same goes for 'normal' attribute).
Repo has 1.658, right?
Apt begs to differ, though.

**Comment by daften** [comment]
That might be a problem with the debian/ubunte repo, i'm not sure. i'd advice you to dig deeper, because on centos, this works perfectly atm.

**Comment by IevgenKabanets** [comment]
Yes, exactly my point... Combining solutions is not the case here.
 Anyway, looking forward your PR merged.

**Comment by CharlesTBetz** [comment]
This wrapper worked for me today on Ubuntu 14.04 (via Vagrant): 
````
node.set[:jenkins][:master][:repository] = 'http://pkg.jenkins-ci.org/debian-stable'
node.set[:jenkins][:master][:repository_key] = 'http://pkg.jenkins-ci.org/debian-stable/jenkins-ci.org.key'
node.override['jenkins']['master']['version'] = '1.651.*'
include_recipe "jenkins::master"
````
My experience was that without specifying debian-stable repo, apt-policy shows only 2.8 available and the attempted 1.651.2 override fails. 

`STDERR: E: Version '1.651.2' for 'jenkins' was not found`

````
apt-cache policy jenkins
jenkins:
  Installed: (none)
  Candidate: 2.8
  Version table:
     2.8 0
        500 http://pkg.jenkins-ci.org/debian/ binary/ Packages
````

With only setting debian-stable, the receipe defaults to the latest available as indicated by @IevgenKabanets.  However, with debian-stable, apt-policy showed 1.651.2 (today) also available:

````
apt-cache policy jenkins
jenkins:
  Installed: 2.7
  Candidate: 2.7
  Version table:
 *** 2.7 0
        100 /var/lib/dpkg/status
     1.651.2 0
        500 http://pkg.jenkins-ci.org/debian-stable/ binary/ Packages
````

And overriding to *exactly* 1.651.2 worked, as did specifying 1.651.* .

I view this as a workaround pending the 2.0 merge (https://github.com/chef-cookbooks/jenkins/pull/471).

**Comment by cheeseplus** [comment]
Closing this out as #557 has been merged along with a few other related fixes.
